Feature/lansastack

robe070 opened the pull request:
Add a template to Autoscale a LANSA Windows VM Scale Set with Azure SQL Database

azurecla commented:
Hi __@robe070__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@robe070__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

robe070 commented:
Do you have a timeframe in which this Pull Request might be reviewed?

singhkays commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,858 @@
+         "type": "Microsoft.Network/virtualNetworks",
+         "name": "[variables('virtualNetworkName')]",
+         "location": "[resourceGroup().location]",
+         "apiVersion": "[variables('networkApiVersion')]",
```
Please use the actual api version instead of variables. This helps with intellisense.

singhkays commented:
@gbowerman @gatneil This is VMSS specific. Could you review this as well?

gatneil commented:
Quick question: we are moving away from WAD/LAD metrics in favor of host metrics (at least for basic metrics). I would feel bad to have this template be immediately out of date. Could you use host metrics instead? The difference is pretty minimal: remove the WAD XML config and use a different metric name. Example here: https://github.com/Azure/azure-quickstart-templates/blob/master/201-vmss-bottle-autoscale/azuredeploy.json

robe070 commented:
Yes we can do that. We'll get it done on Monday. 

Warm regards

Rob Goodridge

On 4 Nov 2016, at 1:54 PM, Neil Sant Gat <notifications@github.com> wrote:

Quick question: we are moving away from WAD/LAD metrics in favor of host metrics (at least for basic metrics). I would feel bad to have this template be immediately out of date. Could you use host metrics instead? The difference is pretty minimal: remove the WAD XML config and use a different metric name. Example here: https://github.com/Azure/azure-quickstart-templates/blob/master/201-vmss-bottle-autoscale/azuredeploy.json

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

robe070 commented:
@gatneil I'd like to understand more about this change to the diagnostics. In particular a list of the available metric names. I have found this: https://azure.microsoft.com/en-us/documentation/articles/insights-autoscale-common-metrics/#compute-metrics-for-linux-vm-v2-as-a-guest-os
but it does not list the metric name your example uses, i.e. "Percentage CPU".

Your example also does not use the "Microsoft.Insights.VMDiagnosticsSettings" extension. I presume it relies on the VM automatically enabling diagnostics. Will it ALWAYS be enabled? There isn't an account setting or some such that might override that behavior?

The documentation I can locate all seems to be discussing WAD/LAD metrics.

And as its a Windows VM, I need the documentation for Windows not Linux.

gatneil commented:
This relies on metrics gathered from the host, not the guest. This is always on. Don't need WAD/LAD Documentation here: https://azure.microsoft.com/en-us/documentation/articles/monitoring-supported-metrics/#microsoftcomputevirtualmachinescalesets. Works across Windows and Linux :)

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+   "parameters": {
+      "msiUri": {
+         "type": "string",
+         "defaultValue": "https://lansalpcmsdn.blob.core.windows.net/app/test/AWAMAPP_v14.1.2_en-us.msi",
```
can we include the MSI with this template instead of having it in a custom storage account? If so, please use _artifactsLocation and _artifactsLocationSasToken as described in the best practice checklist: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

robe070 replied:
The MSI is unique for each deployment. It is created using our Visual LANSA IDE SKU. I think its appropriate for it to be in a custom storage account. The deployment of the MSI is the whole reason for the template in the first place. The template exists to host the application embodied by the MSI.

gatneil replied:
Ok

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+    "sqlserverName": "[concat(variables('namingInfix'), 'sqlserver')]",
+    "gitRepo": "https://raw.githubusercontent.com/robe070/cookbooks/",
+    "gitRefreshName": "git-pull.ps1",
+    "gitRefreshUri": "[concat(variables('gitRepo'), parameters('gitBranch'), '/scripts/', variables('gitRefreshName'))]"
```
Could we include these with the submission instead of pointing to a remote git repo?

robe070 replied:
The git repo is shared between all our deployment environments - AWS, Azure, Google and on-premise. We need to keep them in sync with the Build environment, not with the deployment environment. So we would rather leave it as it is.

gatneil replied:
Ok

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+               "storageProfile": {
+                  "osDisk": {
+                    "vhdContainers": [
+                      "[concat('https://', variables('uniqueStringArray')[0], variables('newStorageAccountSuffix'), '.blob.core.windows.net/', variables('vhdContainerName'))]",
```
please see the best practice around "all cloud" uris: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

robe070 replied:
Scratch that. I've used Azure Resource Explorer to work it out :)

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+                           "forceUpdateTag": "[concat( parameters('installMSI'),parameters('updateMSI'),parameters('triggerWebConfig'),parameters('UninstallMSI'),parameters('fixLicense'))]",
+                           "settings": {
+                              "fileUris": [ "[variables('gitRefreshUri')]" ],
+                              "commandToExecute": "[concat('powershell -NoProfile -ExecutionPolicy unrestricted -command \"& {pushd;./',variables('gitRefreshName'), ' ', parameters('gitBranch'), ';popd;', 'pushd;c:\\lansa\\scripts\\azure-custom-script.ps1 ', '-server_name \"', reference(variables('sqlserverName')).fullyQualifiedDomainName, '\"', ' -DBUT \"SQLAZURE\"',  ' -dbname \"', parameters(' ...
```
this contains passwords, so it should go into the "protectedSettings" instead of "settings"

robe070 replied:
I presume that will be a pain for debugging. Currently I can copy the script from the log and re-execute it for troubleshooting. I presume now the whole lot will be obscured? Is this absolutely mandatory?

gatneil replied:
Strongly recommended. Alternatively, you could call out at the top of the README in big bold letters that this template passes passwords in plain text and thus isn't ready for production deployments (considered a security hole, which is why we strongly recommend using protectedSettings)

robe070 replied:
I expected that :) Just trying to make my dev environment not require even more differences to production. Done.

robe070 replied:
Ok. I applied `protectedSettings` and the deployment failed because the file in `fileUris` was not downloaded:
```
[1048+00000001] [11/09/2016 06:10:18.02] [INFO] Downloading files specified in configuration...
[1048+00000001] [11/09/2016 06:10:18.02] [INFO] No public configuration was provided, no files will be downloaded. Asynchronously executing command: 'SecureCommand_0'

```
Any ideas how to address this?

robe070 replied:
So can we report this as a defect and carry on with this PR as-is?

gatneil replied:
I would recommend debugging further, but if you want we can still call out at the top of the README in big bold letters that this template passes passwords in plain text and thus isn't ready for production deployments

robe070 replied:
This code:
```
                           "publisher": "Microsoft.Azure.Extensions",
                           "type": "CustomScript",
                           "typeHandlerVersion": "2.0",

```
Gets this error:
```
        "message": "Handler 'Microsoft.Azure.Extensions.CustomScript' has reported failure for VM Extension 'FirstInstall' with terminal error 
code '1007' and error message: 'Install failed for plugin (name: Microsoft.Azure.Extensions.CustomScript, version 2.0.2) with exception The 
specified executable is not a valid application for this OS platform.'"
```
Is this a Linux extension? I'm using Windows.

robe070 replied:
I've fixed this by only putting `commandToExecute` in `protectedSettings`

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+                     {
+                        "name": "FirstInstall",
+                        "properties": {
+                           "publisher": "Microsoft.Compute",
```
Instead of CustomScriptForLinux 1.8, Please use: Microsoft.Azure.Extensions CustomScript 2.0

robe070 replied:
I think you mean instead of CustomScriptExtension 1.8 use 2.0. That is, just change the version number?

gatneil replied:
Along with the version bump came a change in publisher and type :)

robe070 replied:
There isn't a version 2.0 of 'Microsoft.Compute', type 'CustomScriptExtension'
```
New-AzureRmResourceGroupDeployment : 11:43:59 AM - Resource Microsoft.Compute/virtualMachineScaleSets 'dbvmsslan' failed with message '{
  "error": {
    "code": "ArtifactNotFound",
    "message": "Extension with publisher 'Microsoft.Compute', type 'CustomScriptExtension', and type handler version '2.0' could not be found in 
the extension repository."
```

robe070 replied:
And I've now checked it against the list of extension versions you sent and the latest is 1.8

gatneil replied:
Publisher: Microsoft.Azure.Extensions
Type: CustomScript
Versin: 2.0 (technically 2.0.2)

robe070 replied:
This code:
```
                           "publisher": "Microsoft.Azure.Extensions",
                           "type": "CustomScript",
                           "typeHandlerVersion": "2.0",

```
Gets this error:
```
        "message": "Handler 'Microsoft.Azure.Extensions.CustomScript' has reported failure for VM Extension 'FirstInstall' with terminal error 
code '1007' and error message: 'Install failed for plugin (name: Microsoft.Azure.Extensions.CustomScript, version 2.0.2) with exception The 
specified executable is not a valid application for this OS platform.'"
```
Is this a Linux extension? I'm using Windows.

For Windows this is what I find: [Windows Custom Script Doc](https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-windows-extensions-customscript/)

gatneil replied:
Oops; you're right; CustomScript is linux-only; my bad; leaving what you already have is fine then

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+                  "name": "Profile1",
+                  "capacity": {
+                     "minimum": "1",
+                     "maximum": "1",
```
if the min and max are 1, why have autoscale?

robe070 replied:
There must be 1 and only 1 of these instances. I'm using the scale set to ensure that if it dies another is created.

gatneil replied:
So, you can have a scale set without autoscale. Adding autoscale with min=max does nothing except make the template more confusing. If the VM host dies, the scale set will automatically move the VM to a new host without autoscale.

gatneil replied:
http://armtg.azurewebsites.net/

robe070 replied:
Excellent resource. Thank you.

robe070 replied:
So can we report this as a defect and carry on with this PR as-is?

gatneil replied:
If the VM fails the load balancer probe, then the load balancer will no longer forward round-robin traffic (in your case, port 80) to that VM, but the VM will stay around. You can still try to use the NAT rules to connect directly to the VM, however (depending on what exactly failed, you may still be able to RDP into the VM even if the web server isn't responding, for instance).

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+               "storageProfile": {
+                  "osDisk": {
+                     "vhdContainers": [
+                        "[concat('https://', variables('uniqueStringArray')[0], variables('newStorageAccountSuffix'), '.blob.core.windows.net/', variables('vhdContainerName'))]",
```
all clouds uri best practice

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+                     {
+                        "name": "Microsoft.Compute.CustomScriptExtension",
+                        "properties": {
+                           "publisher": "Microsoft.Compute",
```
same

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+                           "typeHandlerVersion": "1.8",
+                           "autoUpgradeMinorVersion": "true",
+                           "forceUpdateTag": "[concat( parameters('installMSI'),parameters('updateMSI'),parameters('triggerWebConfig'),parameters('UninstallMSI'),parameters('fixLicense'))]",
+                           "settings": {
```
same

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+         "tags": {
+            "displayName": "SqlServer"
+         },
+         "apiVersion": "2014-04-01-preview",
```
is there a non-preview api version we can use?

robe070 replied:
I've located a recently updated template which has provided the api version.

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,829 @@
+               "tags": {
+                  "displayName": "Database"
+               },
+               "apiVersion": "2014-04-01-preview",
```
same

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,81 @@
+         "value": "lansa-scalable-license"
+      },
+      "vmssName": {
+         "value": "vmsslansa"
```
should probably use GEN-UNIQUE here

robe070 replied:
Will do. And whilst we are on the subject of these placeholders, when I execute the script which is in the PR to create the stack, I need to manually replace these placeholders. With GEN-PASSWORD, that is the literal which is passed to the stack. Not very useful. Is there someway to have that script prompt for the replacements?

gatneil replied:
There are some ps scripts in the root of this repo, which I believe do replace these placeholders. The CI definitely replaces them. For debugging I usually keep around a temporary version with my temporary values for the placeholders

gatneil commented:
Please address the above comments

robe070 commented:
I have addressed all the concerns apart from `protectedSettings` as discussed above.

robe070 commented:
@gatneil I believe I've addresses all the issues. Just the one clarification that the custom script extension you wanted me to use is actually only compatible with Linux.

gatneil commented on lansa-vmss-windows-autoscale-sql-database/azuredeploy.json:
```diff
@@ -0,0 +1,809 @@
+{
+   "$schema": "http://schema.management.azure.com/schemas/2015-01-01-preview/deploymentTemplate.json",
```
would be best to just be 2015-01-01 (without the "-preview"); sorry I missed this before

robe070 replied:
I wondered...I should have asked

gatneil commented:
One last small change around the "$schema" version; after that, it looks good to merge :)

gatneil commented:
Looks good; merging :)

robe070 commented:
Thanks @gatneil. This is my first PR. Is it normal to delete the branch or is it kept open for more changes for the same template?

Also, how long does it usually take for the [website](https://azure.microsoft.com/en-us/documentation/templates/?term=lansa) to be updated? 2 hours later I can see the summary but I get a 404 on the detail page.

gatneil commented:
Looks like it's on the website now :).

With respect to your branch:

Once the PR is merged in the the Azure fork, nothing you do in your fork will affect the Azure fork (unless you open another PR). Do note that over time your fork can diverge from the Azure fork as people make PRs into the Azure fork. To remedy this, I generally use the following work pattern (which, of course, you can adjust as you see fit):

When setting up the repo, I made the Azure fork my "upstream" (short description here: http://stackoverflow.com/questions/8948803/what-does-git-remote-add-upstream-help-achieve)

> When beginning new work, in the master branch of my fork, I do `git pull upstream master` and deal with any merge conflicts (there shouldn't be any in this workflow though). This brings my local fork up to date with Azure master.
> Next, I do a `git checkout -b <branch_name>` and do all the work on this branch.
> Then I do a `git push origin <branch_name>`, which updates my remote fork.
> Do the PR from my remote fork into the Azure master fork
> Once the PR is merged into the Azure master fork, I delete the branch from my local and remote fork.
> Repeat

Obviously you can customize this to your liking, but I think this is a pretty standard workflow.

robe070 commented:
Yes thats the workflow I've done up until now and I was about to delete the branches. I hadn't appreciated that once the PR is closed the link between the repos is stopped. Thanks
